### ISSUE TITLE
[compiler] Distribute labs/compiler decoupled from any specific toolchain

### ISSUE THREAD

**Issue by AndrewJakubowicz** [lit/lit#4350]
I would like to discuss the possibilities of distributing this in other ways than a TS-transform and rollup plugin (as mentioned in the RFC). Not having an option to transform a single file without additional tooling makes it hard or impractical to use the compiler in esoteric or custom build environments. The immediate issue is that it is not possible to use the compiler with esbuild and in extension Vite in development mode.  

It seems to me that the compiler should be standalone and not force consumers to use a certain toolchain that depends on TSC. What would the practical implications be if this was to be implemented without the requirement of TSC?

Concerning providing plugins for other tools, one alternative would be to use some thing like https://github.com/unjs/unplugin. Seems like it should be fairly simple to build given that a transform function for a single file is available.

_Originally posted by @filimon-danopoulos-stratsys in https://github.com/lit/lit/discussions/4117#discussioncomment-7402453_

**Comment by AndrewJakubowicz** [comment]
This is an excellent idea! It could also be a great contribution!

The `labs/compiler` tests already test both TypeScript and JavaScript input files, and do the single file transform via TypeScript's `ts.transpileModule` method: https://github.com/lit/lit/blob/main/packages/labs/compiler/src/test/golden_compiler_test.ts#L65-L71

This single file transpile implementation could be exported, wrapped as a function implementing the `rollup` transform API (which would then be compatible with `unplugin`). Source maps can be retrieved from `ts.transpileModule` by passing in `sourceMap: true` into the `compilerOptions`.

The pro of this approach is it's easy given what we have. The con, is that we also run all the other TypeScript transforms on the provided JavaScript source code. We may need to adjust the compiler settings for JavaScript files so only the templates get transformed, and nothing else.

Is there a set of `compilerOptions` that we can give `ts.transpileModule` that won't change anything else? Does it matter if some stuff minimally is changed?
How is JavaScript using different modules handled (`commonjs`, `amd`, `umd`)?

**Comment by filimon-danopoulos-stratsys** [comment]
The main drawback I can see (if I understand it correctly) with the proposed approach is that the Lit-compiler would essentially act as the TS-compiler for the affected files. Instinctively this seems odd. I would preferably have a solution that accepts either JS or TS and outputs the same language with only the relevant Lit optimizations applied. 

That way any third part tool would essentially still be in control of how to transpilation is performed. 

I don't have enough experience with the TS APIs used to know all the details but at a glance it seems mostly straight forward to use something like the `@babel` family of packages and reuse a lot of the code that already exists. Only reason I mention them is that I have worker a bit with them before. Unfortunately not enough to be able to make an objective comparison.

**Comment by filimon-danopoulos** [comment]
Same Filimon answering, this is just my private account.

Running the compiler in the way you suggested above seems really expensive if applied to every single file in a project. Probably enough to negate the use of faster tools, like `esbuild`. 

Missed that you mentioned module types, I think it is fair to restrict the compiler to modern code, so code written with ES module syntax. This would also mean than any compilation would have to be performed before additional transpilation happens (or simultaneously). I don't see any benefit to allowing other module types.

I have been reading up a bit on the tools provided with typescript and I think it would pretty easy to just change what API:s are used and still keep a lot (probably all) of the same AST-related code. Granted it would require substantial changes to the code overall so before doing anything I would like some inputs on goals and limitations. 

What was the driving factor of creating the compiler as a TS transform rather than a function that applies the compilation? Would it be in the spirit of compiler project to change this?

**Comment by AndrewJakubowicz** [comment]
> I don't see any benefit to allowing other module types.

Agree

> What was the driving factor of creating the compiler as a TS transform rather than a function that applies the compilation?

The driving factor was to target a known "first user" of the compiler (internal Google code). This was to target a known user who can benefit from the work. This also gave a large test codebase to be able to ensure the compiler is working as intended. Internally, there is an easy way to add a TypeScript transform to TypeScript compilation so that shaped the initial public API.

The other hope was that TypeScript would make it easier to add custom transformers natively. This would still only target TypeScript users, but it would avoid the issues with toolchains. https://github.com/microsoft/TypeScript/issues/54276

Our team is also familiar with the TypeScript API, and it's convenient that it handles things like source maps.

> Would it be in the spirit of compiler project to change this?

I think so. Although in my mind we'd probably end up extending it. The current implementation already conforms to the following constraints (which would make writing a JavaScript (ESTree) version easier):
 - We do not allow using the TypeScript compiler to perform type checking based transforms. It's syntactic only.
 - The transform must work at the file-level. There is no way to look across files.

I would still like to expose a TypeScript transform, but we could have a second [ESTree based](https://github.com/estree/estree) transform backend which provides a generic function for applying the Lit compiler transformation to a JavaScript file. The test suite could be shared as well between backends.

> I don't have enough experience with the TS APIs used to know all the details but at a glance it seems mostly straight forward to use something like the @babel family of packages and reuse a lot of the code that already exists

Agree. The general structure of [the compiler currently](https://github.com/lit/lit/blob/main/packages/labs/compiler/src/lib/template-transform.ts) is:

 1. Iterate over the whole file's AST and [find Lit templates](https://github.com/lit/lit/blob/main/packages/labs/compiler/src/lib/template-transform.ts#L185-L216). Found templates get remembered for the next pass.
 2. Iterate over the whole file's AST again, but now rewrite these located templates to the compiled template. This code is mostly agnostic to the AST, it uses [`parse5` to implement `lit-html`'s prepare render phase](https://github.com/lit/lit/blob/main/packages/labs/compiler/src/lib/template-transform.ts#L256).

I need to think more deeply about where would be a good abstraction boundary which could allow for both node types, but from a quick search there [seems to be a mapping between TypeScript and ESTree nodes](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts). 


tl;dr: Currently I'd love the TypeScript transformer export to remain, and I think we could also expose a JavaScript solution by extending the codebase to also traverse and analyze an ESTree based AST.

**Comment by filimon-danopoulos** [comment]
Keeping the transform seems reasonable, there is no objection in that regard from me. 

Why would switching between the TS AST and an ESTree AST be required? I think the tools from the `typescript` package would work fine.

I have had a bit more time to read the code  after familiarizing myself with the TS API. If I am not mistaken what I was initially interested in would be a simple extension to the existing API.  I have a hard time putting this into words so here is some code to illustrate what I am after. 

```js
function optimize(fileName: string, code: string): string {
  const file = ts.createSourceFile(fileName, code)
  const result = ts.transform(file, CompiledTemplatePass.getTransformer()
  const printer = ts.createPrinter();
  
  return printer.printNode(
    ts.EmitHint.SourceFile,
    result.transformed[0],
    ts.createSourceFile('', '')
  )
}
``` 

The approach above seems to handle JS, I did a small verification in isolation with a rudimentary transform

```js
import ts from 'typescript';

const code = 'const foo = html`<div>Test</div>;';
const file = ts.createSourceFile('test.js', code);

const transformFactory = context => root => {
  const visit = node => {
    const currentNode = ts.visitEachChild(node, visit, context);
    if (ts.isIdentifier(node) && currentNode.text === 'foo') {
      return context.factory.createIdentifier('bar');
    }
    return currentNode;
  };
  return ts.visitNode(root, visit);
};

const result = ts.transform(file, [transformFactory]);
const printer = ts.createPrinter();
const transformedSource = printer.printNode(
    ts.EmitHint.Unspecified,
    result.transformed[0],
    ts.createSourceFile('', '')
  )
  
console.log(transformedSource ); // prints: const bar = html `<div>Test</div>;`;
```

I would be in favor of a very simple addition in the spirit of the code above. It does require a `typescript` dependency even for non-TS users but any other solution would require another dependency. Something like proposed solution does not however depend on any `tsconfig.json`.

Sorry for not doing enough research before hand, seems I spoke to early and wasted a bit of time.

**Comment by AndrewJakubowicz** [comment]
I like you `optimize` example, and is along the lines of what I was thinking about regarding using `ts.transpileModule(...)`.

Which could look like:

```js
function optimize(_fileName, source) {
  const result = ts.transpileModule(source, {
      compilerOptions: {
        target: ts.ScriptTarget.Latest,
        module: ts.ModuleKind.NodeNext,
      },
      transformers: {before: [compileLitTemplates()]},
    });
    return {code: result.outputText, map: result.sourceMapText};
}
```

The primary benefit and why I initially thought about this API is that it handles source maps. The negative is that it has less fine grained control over the transforms applied than your example.

I bet there is a way to manually add source maps to your better example. Thank you for brainstorming with me.

**Comment by filimon-danopoulos** [comment]
It is possible to pass compiler options to `.transform` , including `sourceMap: true` but I don't see am obvious way to extract any source maps. 

I guess the transforms are simple enough that creating the sourcemaps manually is an option but it does require substantial additional effort. Additionally it scales quite poorly with additional complexity in compilation. I don't think I can be of much help in figuring out how this is best done as I don't really have a good grasp of the tooling and I can't find any documentation apart from the source.

**Comment by thepassle** [comment]
Whats a little unfortunate about using `transpileModule` is that is also does _other_ transformations. Ideally I'd have some kind of function, like `litCompiler(sourcecode)` that I can run on a string of source code, and have it only do the lit-relevant transforms.

**Comment by filimon-danopoulos** [comment]
Another option to the `ts.traslate` route would be to to use something like https://www.npmjs.com/package/magic-string

The AST traversal could then identify what segments of the text to alter but the actual altering of the text would be done through `magic-string`. The current set of transforms are simple enough that this seems an easier route than generating source maps manually. 

I am pressed on time for this but I could attempt a POC based on this approach if it is of interest. Is there any particular policy regarding third party tooling to consider?

**Comment by filimon-danopoulos** [comment]
I have played around with `magic-string` as an alternative to do the transforms but it creates more problems than it solves in compared to just building the source maps manually. I got a bit more acquainted with the transforms the compiler makes and they are pretty straight forward to create the soucemaps for. 

Additionally, creating the source maps manually allows all code in regards to the transforms to remain the same. I can maybe attempt a PR if you are interested.

**Comment by AndrewJakubowicz** [comment]
Thank you for investigating. I am interested!

It sounds like this would give us a path forward to provide the [`optimize`](https://github.com/lit/lit/issues/4350#issuecomment-1796179136) function you proposed, whilst also exporting source maps! That seems like a clear win!
